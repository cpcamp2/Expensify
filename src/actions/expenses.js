import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    // This is a readable way of setting up default data for expenseData
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    // Set up variable for cleaner code to pass in below. Get rid of uuid since firebase
    // will create unique key for us.
    const expense = { description, note, amount, createdAt };

    // Push expense to firebase database. Pass in ref to then for access to unique key
    // Add return so that I can add another then to chain promises
    // * since this has a promise attached the return is necessary so that
    // another promise can be chained in the test case
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
