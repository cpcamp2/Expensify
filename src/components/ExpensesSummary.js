import React from 'react';
import { connect } from 'react-redux';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
  const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';

  return (
    <div>
      <h4>
        Viewing {props.expenses.length} {expenseWord} totaling {numeral((getExpensesTotal(props.expenses)) / 100).format('$0,0.00')}
      </h4>
    </div>
  );
};


const mapStateToProps = (state) => ({ expenses: state.expenses });

export default connect(mapStateToProps)(ExpensesSummary);
