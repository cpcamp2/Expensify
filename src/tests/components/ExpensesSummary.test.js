import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should correctly render ExpensesSummary component with expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary component without expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]}/>);
  expect(wrapper).toMatchSnapshot();
});
