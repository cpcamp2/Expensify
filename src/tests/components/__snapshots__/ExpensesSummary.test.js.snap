// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary component with expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     expenses totaling 
    $1,141.95
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary component without expenses 1`] = `
<div>
  <p>
    Viewing 
    0
     expenses totaling 
    $0.00
  </p>
</div>
`;
